core/utils/metering.h

```diff
@@ -0,0 +1,247 @@
+
+  Error Add(const T &val, const MeteringKey &key) {
+    Error err;
+    const void *Key_t = (const void *)&key;
+    T *val_t = new T(val);
+    int ret = table_->insert_dpdk(Key_t, val_t);
+    if (!ret) {
+      return MakeError(ENOENT, "Dpdk Insert Failed");
+    }
+    return MakeError(0);
+  }
+
+  Error Delete(const MeteringKey &key) {
+    Error err;
+    bool ret = table_->Remove(key, MeteringKeyHash(total_key_size_),
+                              MeteringKeyEq(total_key_size_));
+    if (!ret) {
+      return MakeError(ENOENT, "rule doesn't exist");
+    }
+    return MakeError(0);
+  }
+
+  void Clear() { table_->Clear(); }
+  int Count() { return table_->Count(); }
+  void DeInit() { table_->DeInit(); }
+
+  // Find an entry in the table.
+  // Returns the value if `key` matches a rule, otherwise `default_value`.
+  T Find(const MeteringKey &key, const T &default_value) const {
+    const auto &table = table_;
+    void *data = nullptr;
+    table->find_dpdk(&key, &data);
+    if (data) {
+      T data_t = *((T *)data);
+      return data_t;
+    } else
+
+      return default_value;
+  }
+
+  uint64_t Find(MeteringKey *keys, T **vals, int n) {
+    uint64_t hit_mask = 0;
+
+    const auto &table = table_;
+    MeteringKey *key_ptr[n];
+    for (int h = 0; h < n; h++)
+      key_ptr[h] = &keys[h];
+    table->lookup_bulk_data((const void **)&key_ptr, n, &hit_mask,
+                            (void **)vals);
+
+    return hit_mask;
+  }
+
+  uint32_t Total_key_size() const { return total_key_size_; }
+
+  void Init(int size) {
+    std::ostringstream address;
+    total_key_size_ = size;
+    address << &table_;
+    std::string name = "Metering" + address.str();
+    dpdk_params.name = name.c_str();
+    dpdk_params.key_len = size;
+    table_.reset(new CuckooMap<MeteringKey, T, MeteringKeyHash, MeteringKeyEq>(
+        0, 0, &dpdk_params));
+  }
+
+  int Srtcm_profile_config(struct rte_meter_srtcm_profile *p,
+                           struct rte_meter_srtcm_params *params) {
+    return rte_meter_srtcm_profile_config(p, params);
+  }
+
+  int Srtcm_config(struct rte_meter_srtcm *m,
+                   struct rte_meter_srtcm_profile *p) {
+    return rte_meter_srtcm_config(m, p);
+  }
+
+  int Trtcm_profile_config(struct rte_meter_trtcm_profile *p,
+                           struct rte_meter_trtcm_params *params) {
+    return rte_meter_trtcm_profile_config(p, params);
+  }
+
+  int Trtcm_config(struct rte_meter_trtcm *m,
+                   struct rte_meter_trtcm_profile *p) {
+    return rte_meter_trtcm_config(m, p);
+  }
+
+  int Trtcm_rfc4115_profile_config(
+      struct rte_meter_trtcm_rfc4115_profile *p,
+      struct rte_meter_trtcm_rfc4115_params *params) {
+    return rte_meter_trtcm_rfc4115_profile_config(p, params);
+  }
+
+  int Trtcm_rfc4115_config(struct rte_meter_trtcm_rfc4115 *m,
+                           struct rte_meter_trtcm_rfc4115_profile *p) {
+    return rte_meter_trtcm_rfc4115_config(m, p);
+  }
```

Comment:
Are these used anywhere?

Reply:
lets keep it, in future if endor want to use any algo they shall be able to do it by calling these avaialble Api's without doing any major changes in architecture.